File: core/java/src/main/java/hara/lang/protocol/IApplicable.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Context-aware application and transformation protocol. */
@HaraProtocolBinding(namespace = "std.protocol.iapplicable", name = "IApplicable")
public interface IApplicable {
  @HaraMethod(value = "apply-in", arity = 3)
  Object applyIn(Object runtime, Object[] args);

  @HaraMethod(value = "apply-default", arity = 1)
  default Object applyDefault() {
    return this;
  }

  @HaraMethod(value = "transform-in", arity = 3)
  Object transformIn(Object runtime, Object[] args);

  @HaraMethod(value = "transform-out", arity = 4)
  Object transformOut(Object runtime, Object[] args, Object value);
}

File: core/rust/src/lang/protocol/iapplicable.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.iapplicable", name = "IApplicable")]
pub trait IApplicable<R, A> {
    type Output;

    #[hara_method(value = "apply-in", arity = 3)]
    fn apply_in(&self, runtime: &mut R, arguments: A) -> Self::Output;
    #[hara_method(value = "apply-default", arity = 1)]
    fn apply_default(&mut self) -> &mut R;
    #[hara_method(value = "transform-in", arity = 3)]
    fn transform_in(&self, runtime: &R, arguments: A) -> A;
    #[hara_method(value = "transform-out", arity = 4)]
    fn transform_out(&self, runtime: &R, arguments: A, value: Self::Output) -> Self::Output;
}
